File: eon_timer/timers/frame_timer.py
from typing import Final

from eon_timer.util import const

from .calibrator import Calibrator


class FrameTimer:
    def __init__(self, calibrator: Calibrator):
        self.calibrator: Final = calibrator

    def create(self, pre_timer: int, target_frame: int, calibration: float) -> list[float]:
        return [pre_timer, self.create_phase(target_frame, calibration)]

    def create_phase(self, target_frame: int, calibration: float) -> float:
        return self.calibrator.to_milliseconds(target_frame) + calibration

    def calibrate(self, target_frame: int, frame_hit: int) -> float:
        return self.calibrator.to_milliseconds(target_frame - frame_hit)


class VariableFrameTimer:
    def __init__(self, frame_timer: FrameTimer):
        self.frame_timer: Final = frame_timer

    def create(self, pre_timer: int) -> list[float]:
        return [pre_timer, const.INFINITY]

    def calibrate(self, target_frame: int, frame_hit: int) -> float:
        return self.frame_timer.calibrate(target_frame, frame_hit)

File: src/timers/frameTimer.ts
import { INFINITY } from '../utils/constants';
import { CalibratorSettings, toMilliseconds } from './calibrator';

export function createFramePhases(
  settings: CalibratorSettings,
  preTimer: number,
  targetFrame: number,
  calibration: number,
): number[] {
  return [preTimer, createFramePhase(settings, targetFrame, calibration)];
}

export function createFramePhase(
  settings: CalibratorSettings,
  targetFrame: number,
  calibration: number,
): number {
  return toMilliseconds(settings, targetFrame) + calibration;
}

export function calibrateFrame(
  settings: CalibratorSettings,
  targetFrame: number,
  frameHit: number,
): number {
  return toMilliseconds(settings, targetFrame - frameHit);
}

export function createVariableFramePhases(preTimer: number): number[] {
  return [preTimer, INFINITY];
}
